sat/data/schema/bootprep_schema.yaml

```diff
@@ -34,6 +51,38 @@ description: >
 type: object
 additionalProperties: false
 properties:
+  schema_version:
+    description: >
+      The version of the bootprep input file schema used by this input file. The
+      version of the schema should be a SemVer-like version number with three
+      components, major.minor.patch. This will be populated with the correct
+      value in input files generated by 'sat bootprep generate-example'.
+
+      This version will be checked against the current schema version understood
+      by the 'sat bootprep' commmand.
+
+      Compatibility is defined as follows. Let us refer to the schema version
+      specified by the bootprep input file as the "input version" and the current
+      schema version used by the 'sat bootprep' command as the "current version".
+
+
+      - If the input version has an older major version than the current
+        version, it is incompatible.
+
+      - If the input version has an older minor version than the current
+        version, it may be incompatible, but will be accepted with a warning.
+
+      - If the input version has an older patch version than the current
+        version, it is compatible.
+
+      - If the input version is newer than the current version, it is
+        incompatible because it could contain new fields not understood by the
+        current version.
```

Comment:
The more I think about this, I'm wondering if we want a similar nuance for when the input version is newer than the current version -- i.e. if just the input patch version is newer, then theoretically the older version of the schema should be compatible, though the validation would probably throw up errors on unrecognized fields. I wonder if there is a option that can be passed to the validator to ignore errors or throw away information from unrecognized fields. This may be outside the scope of this change, and I don't think it's a deal-breaker as is, but I just got to thinking about it more as I looked again.

Reply:
Yeah, good question. Because we have `additionalProperties: false` in most places throughout the schema, I think generally speaking, when we add new properties to the schema, newer input files that use that version are not going to be validate against older schema versions because like you said, there will be errors on unrecognized properties. As far as a way to ignore extra properties, one way would be to stop setting `additionalProperties: false. However, [this article](https://snowplowanalytics.com/blog/2014/05/13/introducing-schemaver-for-semantic-versioning-of-schemas/) I read suggested that disallowing additional properties is a helpful way to ensure that files which validated with older versions of the schema will still validate with the newer version of the schema. See section 2.1 and 2.2 there.

It feels most important to me to support older input files with newer versions of the schema and bootprep code than it is to support newer versions of the input file schema with older versions of the schema and bootprep code. I think this approach handles that more important use case pretty well.

I'll keep this in mind though, and we can discuss further if you'd like.